geopandas/geodataframe.py

```diff
@@ -312,19 +312,27 @@ def set_geometry(self, col, drop=False, inplace=False, crs=None):
             level_crs = getattr(level, "crs", None)
             crs = level_crs if level_crs is not None else self._crs
 
-        if isinstance(level, (GeoSeries, GeometryArray)) and level.crs != crs:
-            # Avoids caching issues/crs sharing issues
-            level = level.copy()
-            level.crs = crs
-
-        # Check that we are using a listlike of geometries
-        level = _ensure_geometry(level, crs=crs)
-        index = frame.index
-        frame[geo_column_name] = level
-        if frame.index is not index and len(frame.index) == len(index):
-            # With pandas < 1.0 and an empty frame (no rows), the index gets reset
-            # to a default RangeIndex -> set back the original index if needed
-            frame.index = index
+        # skip assignment when not needed (sindex preservation)
+        assignment_required = (
+            not isinstance(col, str)
+            or not isinstance(level, GeoSeries)
+            or crs != level.crs
+            or drop
+        )
+        if assignment_required:
+            if isinstance(level, (GeoSeries, GeometryArray)) and level.crs != crs:
+                # Avoids caching issues/crs sharing issues
+                level = level.copy()
+                level.crs = crs
+
+            # Check that we are using a listlike of geometries
+            level = _ensure_geometry(level, crs=crs)
+            index = frame.index
+            frame[geo_column_name] = level
+            if frame.index is not index and len(frame.index) == len(index):
+                # With pandas < 1.0 and an empty frame (no rows), the index gets reset
+                # to a default RangeIndex -> set back the original index if needed
+                frame.index = index
```

Comment:
This was to fix 
```python
df = GeoDataFrame(df)
df._geometry_column_name = geometry
```

being required in the _constructor due to the sindex cache being reset. The line `frame[geo_column_name] = level` causes the id of the sindex to change, I don't know enough about pandas internals to know why this happens, but I could trace it to `_iset_item_mgr` in pandas.core.frame.py.
[Note that I've still failed in being able to remove those lines, as they're also needed so that concat(axis=1) with duplicate geometry columns doesn't have recursion errors. This is probably a bit hard to follow but this is related to the discussion on concat on the old diff. 

```python
pd.concat([gdf, gdf], axis=1)
```
will raise errors in either __init__ or in set_geometry, depending on whether gdf._geometry_column_name is geometry or not. Passing through without a geometry column in the init 
```python
df = GeoDataFrame(df)
df._geometry_column_name = geometry
if crs is not None:
    df.set_crs(crs, inplace=True)
```

skips the error and allows this to be caught in finalise, where an error specific to concat can be raised. I'm not a huge fan of this to be honest given this could occur in other ways and this could go unchecked. 
My preference would actually be to feed geometry and crs to the constructor normally, and perhaps change the duplicate column error to something like:
```
raise ValueError('GeoDataFrame does not support setting the geometry column '
            'where the column name is shared by multiple columns. '
            '(This can occur indirectly from calling pd.concat(..., axis=1))'
```